libs/chonky/src/embedding/multi_modal_embedding.rs

```diff
@@ -0,0 +1,338 @@
+use base64::{Engine as _, engine::general_purpose};
+use error_stack::{Report, ResultExt as _};
+use image::DynamicImage;
+use reqwest::{
+    Client,
+    header::{AUTHORIZATION, CONTENT_TYPE, HeaderMap, HeaderValue},
+};
+use serde_json::{Value, json};
+
+use crate::{
+    ChonkyError, DocumentEmbeddings, Embedding, StructuralEmbedding, StructuralMetadata,
+    pdf_segmentation::ExtractedTable,
+};
+
+fn get_vertex_access_token() -> Result<String, Report<ChonkyError>> {
+    Ok(String::from_utf8(
+        std::process::Command::new("gcloud")
+            .args(["auth", "print-access-token"])
+            .output()
+            .change_context(ChonkyError::VertexAPI)?
+            .stdout,
+    )
+    .change_context(ChonkyError::VertexAPI)?
+    .trim()
+    .to_owned())
+}
+
+fn base64_json(image_data: Vec<u8>) -> Result<String, Report<ChonkyError>> {
+    let base64_encoded_img = general_purpose::STANDARD.encode(image_data);
+    let formatted_json = serde_json::to_string_pretty(&json!({
+        "instances": [
+            {
+                "image": {
+                    "bytesBase64Encoded": base64_encoded_img
+                }
+            }
+        ]
+    }))
+    .change_context(ChonkyError::VertexAPI)?;
+
+    // add newline that is appended by yarn in the snapshot
+    Ok(format!("{formatted_json}\n"))
+}
+
+fn text_json(text: &[String]) -> Result<String, Report<ChonkyError>> {
+    //merge all text into one without seperator for now
+    let mut text = text.join("");
+    text.truncate(1000);
+    let formatted_json = serde_json::to_string_pretty(&json!({
+        "instances": [
+            {
+                "text": text
+            }
+        ]
+    }))
+    .change_context(ChonkyError::VertexAPI)?;
+
+    // add newline that is appended by yarn in the snapshot
+    Ok(format!("{formatted_json}\n"))
+}
+
+/// Given the extracted images from the pdf, embeds them
+///
+/// # Errors
+///
+/// [`ChonkyError::VertexAPI`] when there are HTTP request errors
+pub async fn embed_pdf_object_images(
+    pdf_image_extract: Vec<Vec<DynamicImage>>,
+    project_id: &str,
+) -> Result<Vec<Vec<Vec<f64>>>, Report<ChonkyError>> {
+    let mut embeddings = Vec::new();
+    for page_images in pdf_image_extract {
+        embeddings.push(embed_screenshots(page_images, project_id).await?);
+    }
+    Ok(embeddings)
+}
+
+/// Given the screenshot of each page in pdf return its embeddings
+///
+/// # Errors
+///
+/// [`ChonkyError::VertexAPI`] when there are HTTP request errors
+pub async fn embed_screenshots(
+    pdf_image_extract: Vec<DynamicImage>,
```

Comment:
Since `embed_screenshots` ends up taking ownership of the images in the form of transforming the image to bytes to send to the API embedding request this may not be possible.

Reply:
Ah, yes. At this point I wonder if we want to really take ownership of the image in `embed_screenshots` (and if the image needs to be stored alongside the embedding). My instinct would be no. If you need to have them associated, you could call this function like this:
```rust
let images: Vec<DynamicImage> = todo!();
let embedded_images = Vec<(ImageEmbedding, DynamicImage)> = embed_pdf_object_images(&images)?.zip(images).collect();
```

Reply:
Actually, as you don't need to rely on any slice-operation here it would be better to pass `impl IntoIterator<Item = &DynamicImage>` instead. `IntoIterator` is implemented for any thing which can be converted into an iterator (e.g. vectors, arrays, slices, but also for sets, etc.). That's the typical way to take a list of things in a function and instead of returning a `Vec` you return an `impl Iterator<Item = ImageEmbedding>`. In this case it's a bit unfortunate because we return a `Result` and use `async`, though, so returning a `Vec` here is much simpler.

```suggestion
    pdf_image_extract: impl IntoIterator<Item=&DynamicImage>,
```
or the owned version:
```suggestion
    pdf_image_extract: impl IntoIterator<Item=DynamicImage>,
```

I have not tested this locally, but it's possible that you need to add a few `Send` bounds.

Reply:
The actual idiomatic version (but also quite hard to start with) Is using [streams](https://docs.rs/futures/latest/futures/stream/index.html).
This way the function would become
```rust
pub fn embed_pdf_object_images(
    pdf_image_extract: impl IntoIterator<Item = PageImageObjects>,
    project_id: &str,
) -> impl Stream<Item = Result<PageImageObjectsEmbeddings, Report<ChonkyError>>> {
    stream::iter(pdf_image_extract)
        .then(|page_images| embed_screenshots(page_images, project_id))
        .map(|embeddings| {
            Ok(PageImageObjectsEmbeddings {
                _embeddings: embeddings?,
            })
        })
}
```
Note, that this function is _not_ `async`, because it returns a `Stream` (which is pretty much an async iterator).
I don't say we need to take this route because streams are not yet well matured in Rust (and a more built-in feature is on the way), but it avoids allocations.